src/http/api/resources/files.js

```diff
@@ -87,57 +87,57 @@ exports.get = {
 
   // main route handler which is called after the above `parseArgs`, but only if the args were valid
   handler: (request, reply) => {
-    const key = request.pre.args.key
+    const cid = request.pre.args.key
     const ipfs = request.server.app.ipfs
     const pack = tar.pack()
 
-    ipfs.files.getPull(key, (err, stream) => {
-      if (err) {
-        log.error(err)
+    const stream = ipfs.files.getPullStream(cid)
 
-        reply({
-          Message: 'Failed to get file: ' + err,
-          Code: 0
-        }).code(500)
-        return
-      }
+    pull(
+      stream,
+      pull.asyncMap((file, cb) => {
+        const header = { name: file.path }
 
-      pull(
-        stream,
-        pull.asyncMap((file, cb) => {
-          const header = { name: file.path }
-          if (!file.content) {
-            header.type = 'directory'
-            pack.entry(header)
+        if (file.content) {
+          console.log('packing file:', header.name)
+
+          header.size = file.size
+          const ps = pack.entry(header, () => {
+            console.log('packed a file')
             cb()
+          })
+          // in case request has been aborted
+          if (!ps) {
+            log('other side hung up')
           } else {
-            header.size = file.size
-            const packStream = pack.entry(header, cb)
-            if (!packStream) {
-              // this happens if the request is aborted
-              // we just skip things then
-              log('other side hung up')
-              return cb()
-            }
-            toStream.source(file.content).pipe(packStream)
-          }
-        }),
-        pull.onEnd((err) => {
-          if (err) {
-            log.error(err)
-            pack.emit('error', err)
-            pack.destroy()
-            return
+            toStream.source(file.content).pipe(ps)
           }
+        } else {
+          console.log('packing dir:', header.name)
 
-          pack.finalize()
-        })
-      )
+          header.type = 'directory'
+          pack.entry(header, () => {
+            console.log('packed a directory')
+            cb()
+          })
+        }
```

Comment:
The final issue is here, for some reason, tar-stream blocks on the second dir that is added:

<img width="598" alt="image" src="https://user-images.githubusercontent.com/1211152/33010837-dcac59ba-cdd3-11e7-9886-a312c05ccbb4.png">